### DISCUSSION TITLE
Ratelimit handling & Pagination

### DISCUSSION THREAD

**Discussion by anthonyvancauwenberghe** [saloonphp/saloon#217] [General]
Hi,

For a project these past few months I have been using a simpler, self created implementation of saloon (wish I had known this amazing package existed sooner). I've been looking through the documentation & code (of both v1 & v2), but i'm missing 2 (common) features that I need quite often while working with external api's: 
the first one is pagination handling (for which I see there is already a draft? https://github.com/Sammyjo20/Saloon/pull/152). 
The second is rate limit handling.

Are you open to accepting PR's for rate limit handling? Since I'm new to this package could you give some basic pointers on how/where to start with this.  

Kind regards,
Anthony

**Comment by Sammyjo20** [comment]
Hey @anthonyvancauwenberghe 

Many thanks for your message and your kind words on Saloon, please excuse the missing pages in the v2 docs, I'm still working on them in a flurry!

Ah! Well yes pagination is in progress, @juse-less and I are working on it together for version two. 

Pagination & rate limiting where my two biggest bug bears with APIs too. I've implemented rate limiting with Laravel's jobs before but I'd LOVE to discuss how we could implement it with Saloon. Do you have any ideas around it? It'll be good to have a method we agree on before a PR is made

Many thanks!

**Comment by juse-less** [comment]
@Sammyjo20 Off the top of my head, to stay agnostic, we could possibly use [PSR-6](https://www.php-fig.org/psr/psr-6/), or [PSR-16](https://www.php-fig.org/psr/psr-16/) to store the throttling.
These are caches, but are defined as PSRs, making it more versatile/agnostic than implementing it for specific solutions (like Laravel), or an internal storage, or so.
Then people can provide their PSR-supported caching, like something from Laravel, Symfony, or so.

In terms of implementation, at least external APIs should return with HTTP 429 if you're being throttled.
These also contains datetime/interval for when you can try again.
This datetime/interval usually when the throttle is lifted.
It's also used (in Laravel, for instance) on maintenance mode, at which point it's more of 'try again later' datetime/interval.
But for actual throttling, it should be static, as they'd know when your throttling lifts.

Saloon also already has an Exception for HTTP 429, which might be useful for that.

**Comment by anthonyvancauwenberghe** [comment]
Hi @Sammyjo20,

Happy to hear you're open to PR's!

So in my head the rate limit interface could roughly look something like this:

isThrottled() -> e.g. $response->statusCode === 429
retryAt() -> extract from response (or cache) 
timeout() -> maximum total time to wait before throwing error
retries() -> maximum total requests to execute before throwing error
cacheDriver() -> PSR-16 implementation of a cache interface.
handle() -> store throttling in cache & keep sleeping until retryAt is in the past or timeout/retrylimit is hit.

Let me know what you think.

**Comment by Sammyjo20** [comment]
Thanks both for your replies, all very useful information - I actually implemented PSR-6/16 with the Saloon cache plugin - I made sure people "resolved" the cache driver they want to use, either "LaravelDriver", "SimpleCache" or "Flysystem" drivers. 

I like the ideas around the methods you've tried above, how would you imagine Saloon would actually attempt to retry the request, would that be completely up to the developer (e.g they should queue a job that makes the request again, or maybe use a CRON job)? I see what you're doing with `sleep()` which definitely could work, but then I was also thinking about long wait times, like if you've exhausted all your requests for a day and need to wait 12 hours.

**Comment by anthonyvancauwenberghe** [comment]
By default I would just sleep and retry until success or throw an exception when timeout/retry limit is hit. Personally I would like to catch the exception and requeue if that happens. But developers could always extend and override the default ratelimit class and implement their own behavior?

What do you think?

Kind regards

**Comment by Sammyjo20** [comment]
Yeah that could definitely work! 🔥 If people could handle how it should be retried, then people will just retry with laravel jobs. That sounds amazing!

**Comment by anthonyvancauwenberghe** [comment]
We could include the time for the ratelimit to disappear in the exception class. That way you could e.g. easily catch it in the laravel job class and requeue with a delay equal to the time it takes for the ratelimit to disappear. How would you like to proceed with this?

**Comment by anthonyvancauwenberghe** [comment]
I've been going over the code and think it's mostly quite straight forward to implement. 
The only thing where I can see possible issues, is with the sending of async requests. The default action would be to wait for the ratelimit to disappear and then send the request again. 
If ratelimited we should sleep before the request is sent (check retryAt value from cache), but doing so inside the dispatcher for async requests is not the right way, as it will block the main thread, instead of blocking only that async request. Not sure if there is a way to "delay" execution of async requests?

**Comment by Sammyjo20** [comment]
Could you implement the delay inside of a request middleware? Those are run for both sync and async requests? You could build a `CheckRateLimit` middleware, it could even go inside of the registerDefaultMiddleware method on the pending request and then that will perform the sleep before sending the next request?

**Comment by anthonyvancauwenberghe** [comment]
I havn't looked at middleware yet but if you're able to retrigger requests from the middleware that's definately an option for sync requests. For async requests however it's not an ideal situation as it will delay the whole pool of requests rather than just the single request.

**Comment by Sammyjo20** [comment]
I just found out that Guzzle has this: Would this help? 

https://docs.guzzlephp.org/en/stable/request-options.html#delay

**Comment by anthonyvancauwenberghe** [comment]
Yes it does! I've submitted a PR for this #160

**Comment by anthonyvancauwenberghe** [comment]
I'm currently looking at ratelimiting and how to implement this with middleware. 
However it looks like the way middleware is currently implemented doesn't support resending requests (from the middleware itself). 

In laravel you would be able to call $next() and receive a response. Therefor having 1 middleware for both request & response. Can we do the same for V2 of saloon?

**Comment by Sammyjo20** [comment]
Could you build it as a response middleware? Then inside the response middleware, if it fails you can send another request and return that request’s response?

**Comment by Sammyjo20** [comment]
The response middleware accepts a response object to be returned and you can change it as long as it’s still another response object

**Comment by anthonyvancauwenberghe** [comment]
No you would need 2 middlewares. One (requestmiddleware) to check if the ratelimit is hit (coming from the cache) before the request can be send if it exceeds the treshold an exception needs to be thrown otherwise it needs to wait until the limit has passed. And another (responsemiddleware) that catches the exception after you send the request and somehow did hit the ratelimit to resend the request with a delay (or throw exception).

**Comment by Sammyjo20** [comment]
Unfortunately the two middleware pipelines are separate in v2. I can look to improving it later down the line for v3 but it’s going to be separate for this release. 

I think your idea could work with current middlware.

**Comment by Sammyjo20** [comment]
Hey @anthonyvancauwenberghe with Saloon v2 released, any more thoughts on how you would implement rate limiting? 😀

**Comment by Sammyjo20** [comment]
Hey @anthonyvancauwenberghe! 

I hope you are well. I've had some inspiration tonight and I'm going to work on my own PoC for rate limit handling - it's a little different to your approach - instead when it hits a rate limit it will actually throw an exception where the application should handle the delay/retry. I've also got an idea to use the `delay` option you very kindly built into Saloon too.

Here's a rough idea of how the public-api might look. I've just been riffing with some ideas so far.

```php
<?php

class ForgeConnector extends Connector
{
    use HasRateLimits;

    // Define your rate limits for the API
    
    protected function resolveRateLimits(): array
    {
          return [
                Limit::allow(100)->everyMinute(), // Will release after 60 seconds
                Limit::allow(250)->everyFiveMinutes(), // Will release after 5 minutes
                
                // Below has a threshold of 95% so it will check if the limiter is at that amount
                // if it is, then it will limit for the time it needs
                
                Limit::allow(allow: 1000, threshold: 0.95)->everyDay(), // Will release after 1 day
                
                // Below will actually use Guzzle delay instead of throwing an exception.
                
                Limit::allow(250)->everyFiveMinutes()->useDelay(),
                
                // Customise how the limit is lifted with "releaseAfterSeconds"
                'custom-key-if-you-like' => Limit::allow(1000)->everyDay()->releaseAfterSeconds(60), 
          ];
    }
    
    protected function handleRateLimit(Response $response, Limit $limit)
    {
            if ($response->status() === 429) {
                 $limit->reached(releaseAfterSeconds: 120); // Will throw an exception
            }
            
            // Otherwise increase the limit
            
            $limit->hit();
    }
}
```

My ideas are here:

- Use a request middleware to count the number of requests being sent to an API
- Use a response middleware to look out for any 429 errors and throw an exception
    - We could allow people to customise how Saloon determines a response as rate limited, we need the ability to check the response status or we could check something in the body
    - We should also allow the user to define things like limit remaining and when we can try again etc
- We need somewhere to define the various rate limits, some services have a per-minute limit, some services have per hour, per day etc
- We need somewhere to resolve/define the rate limiting driver
    - Redis
    - Memcache
    - File
- The rate limit driver will be responsible for storing the amount of requests being made (key can be “connector class name + limiter name”) and the value is the amount of requests being made
- When the amount of requests is within the threshold (e.g 90%) then throw exceptions if any more requests are made with a “time remaining”
- We should have a key in the store like “connector class name + ‘release-at’” and then the value could be the unix timestamp of when the requests will be allowed again, we can also work out number of seconds remaining too
- When the exception is thrown you can get the number of seconds remaining so you know to throttle back things
- You could also use helper methods to check if we are approaching a limit like $connector→hasReachedRateLimit()

**Comment by anthonyvancauwenberghe** [comment]
Hi!

I think your approach is a good start. But you are assuming the saloon app is the only service consuming from the external api (Which in my experience is often not the case). A lot of api's add headers that pass the remaining amount of requests left in a certain interval with in addition a retry-after header when the ratelimit is hit. We could define an interface that's responsible for parsing these headers and then adds them to your Limit() method. 

Let me know what you think.
Kind regards.

**Comment by Sammyjo20** [comment]
Hey! 

Thank you - I'm excited to give it a go - I'll let you know when my PoC is ready to look at. Regarding what you said - on the connector I've added a method called "processRateLimit" (renamed from above) and after every request, every limit will be iterated over and you can manually check the response and define if a rate limit has already been hit

```php
protected function processRateLimit(Response $response, Limit $limit)
{
    if ($response->status() === 429) {
        $retryAt = $response->header('Retry-At');

        $limit->reached(releaseAfterSeconds: $retryAt); // Will throw an exception
    }

    // Otherwise increase the limit

    $limit->hit();
}
```

**Comment by Sammyjo20** [comment]
This is my basic PoC I'm just brainstorming of how my middleware will work

```php
// Request middleware can check the rate

$middleware->onRequest(function (PendingRequest $pendingRequest) {
    $driver = $pendingRequest->getConnector()->getRateLimitDriver();
    
    if ($driver->hasBeenLimited()) {
      throw new LimitExceededException($driver->getTimeRemainingInSeconds());
    }
});

// Response middleware could check the rate

$middleware->onResponse(function (Response $response) {
    // Check here if a limit already exists on the driver
	    
    $limits = $driver->resolveLimits(); // Returns array of populated limits
    
    foreach($limits as $limit) {
        // handleRateLimit is on the trait/connector
        
        $limit = tap($limit, fn () => $this->processRateLimit($response, $limit));
        
        if (! $limit->hasReachedLimit()) {
            continue;
        }
        
        // Throw/Delay here
        // Could run something like $this->recordLimit($limit->getReleaseInSeconds())
        
        throw new LimitExceededException($limit->getReleaseInSeconds());
    }
});

// Response middleware could check if an exception is thrown 
```

**Comment by Sammyjo20** [comment]
Hey @anthonyvancauwenberghe!

So I actually built a working PoC and I'm really excited about it.

If you want to have a look at it you can here: https://github.com/Sammyjo20/Saloon/pull/178

I'm really happy I could build it all petty much in one trait - if you want to try it locally you can use the test I added - just make sure Redis is running and it should work :D 

Still lots to do of course but I'm very happy I got this far.

**Comment by epicuristic** [comment]
Will the scenario be supported where an external service returns limiting headers in milliseconds? I wouldn't want to wait for seconds.
Real case headers:
X-Lognex-Reset: 0 - time to reset the limit in milliseconds
X-Lognex-Retry-TimeInterval: 3000 - the interval in milliseconds during which requests can be made
X-RateLimit-Limit: 45 - the total number of queries that can be made evenly during the interval before the 429 error
X-RateLimit-Remaining: 44 - the remaining number of requests that can be sent before receiving 429

More of that, it would be nice to have an ability to force requests to wait until the limit expires rather than getting an exception.

**Comment by Sammyjo20** [comment]
Hey @epicuristic 

Just thinking about your scenario - in this instance would you want to define any limits in the `resolveLimits` method on the connector? Or are you expecting it to look at the headers and work it all out itself? 

I could introduce a special kind of limiter that looks something like: 

```php
Limit::useHeaders(
    allow: 'X-RateLimit-Limit',
    every: 'X-Lognex-Retry-TimeInterval',
)
```

Then inside of your connector you can write something like:

```php
protected function processLimit(Limit $limit, $response)
{
     if ($response->header('X-RateLimit-Remaining') === 0) {
         $limit->exceeded();
     }

     $limit->hit();
}
```

Let me know your thoughts :) 

In terms of milliseconds, so far I've built it using Redis - I think the lowest TTL is one second with Redis, but we could certainly convert the interval that's in milliseconds into seconds.

**Comment by epicuristic** [comment]
I think defining limits inside `processLimit` is more preferable because it is more controllable.
As for milliseconds, as I know, redis has `psetex` that accepts milliseconds.

**Comment by Sammyjo20** [comment]
@epicuristic Still working on the millisecond limits - however I've got a working PoC for what I'm calling "greedy" limits which are definable limits that basically will allow as many requests as you like, but you must tell the limiter when it is out of requests.

Not sold on the name, but here's how it looks. 

Basically, it will allow every request, but as soon as it hits a 429 it will set how many seconds until it should retry, in this instance it'll get the timestamp in the X-Ratelimit-Reset and then work out how many seconds from then.

```php
Limit::greedy(function (Response $response, Limit $limit) {
    if ($response->status() === 429) {
        $resetSeconds = (int)$response->header('x-ratelimit-reset', 0);

        $limit->exceeded($resetSeconds - now()->timestamp);
    }
})
```


Next time you try to make an API call, instead of making a real request, Saloon will check the limiter and see it's still not expired and will throw an exception before the request is made. You can also use

`$connector->hasReachedLimit()` in your queued jobs to automatically release them back onto the queue and check again

**Comment by Sammyjo20** [comment]
Hey @anthonyvancauwenberghe @epicuristic 

Just wanted to let you know that I'm still working on this. I did a small write up on how it works and you can see the code here. It's not finished yet, but thought you might be interested and feel free to leave any feedback:

https://github.com/Sammyjo20/saloon-rate-limiter-plugin
